akka-actor-typed/src/main/scala/akka/actor/typed/BehaviorInterceptor.scala

```diff
@@ -14,30 +17,35 @@ import akka.annotation.{ DoNotInherit, InternalApi }
  * If the interceptor does keep mutable state care must be taken to create the instance in a `setup` block
  * so that a new instance is created per spawned actor rather than shared among actor instance.
  *
+ * @param interceptMessageClass Allows for applying the interceptor only to certain message types.
+ *                              If the message is not of this class or a subclass thereof it will
+ *                              bypass the interceptor and be continue to the inner behavior untouched.
  * @tparam O The outer message type – the type of messages the intercepting behavior will accept
+ * @tparam M The middle message type – the type of messages the interceptor cares about. Allows for
+ *           applying the interceptor only to certain message types.
+ *           If the message is not of this class or a subclass thereof it will
+ *           bypass the interceptor and be continue to the inner behavior untouched.
  * @tparam I The inner message type - the type of message the wrapped behavior accepts
  */
-abstract class BehaviorInterceptor[O, I] {
+abstract class BehaviorInterceptor[O, M <: O, I](val interceptMessageClass: Class[M]) {
```

Comment:
Isn't this still problematic for event sourced where `M` isn't `<: O`?

What about dropping the type params entirely and thinking of interceptor as super low level (which it is), shouldn't that allow us to not have the extra param/classtag everywhere?

Reply:
This works fine for interceptor in EventSourcedBehavior because that would be [Any, Any, InternalProtocol] and an outer interceptor would be [Command, C <: Command, Command]

By dropping the type param, you mean that an interceptor would always have to handle Any? I don’t think that is good for the more specific interceptors such as withMdc.

Reply:
How would a user interceptor know that it is `Any`, they think it is `TheirCommand`

I was thinking that it would push the decision if a `Class`-instance was needed to the specific interceptor where it actually was needed.

Reply:
I think there may be something more hiding here, `intercept(Interceptor[O, M, I])(Behavior[I])` turns `Behavior[I]` to a `Behavior[O]` but it only actually changes `M` which means that unless `M == O` or the single subtype there is a set of messages that will be passed to the inner behavior and potentially cause a classcast there?

Reply:
You are right. `_ < O` was wrong thinking, and therefore also the third type parameter. I have reverted that so now an interceptor must handle all `O`.